### PR TITLE
PLAT-566 Fixed Transaction, Validation and UI Handling in EMF Forms

### PR DESCRIPTION
**Pull request by oliver-richers**

### CODE REVIEW

File: platform-db-core/src/main/java/com/softicar/platform/db/core/transaction/IDbTransaction.java [hunk 1]
```diff
@@ -1,16 +1,61 @@
...
+import com.softicar.platform.db.core.connection.IDbConnection;
 import java.util.Optional;
 import java.util.function.Supplier;
 
+/**
+ * An {@link IDbTransaction} is an object representing a logical transaction on
+ * the database management system (DBMS).
+ * <p>
+ * An {@link IDbTransaction} may represent a root (physical) transaction or a
+ * nested (logical) transaction.
```
Comment (alexander-schmidt) [inline platform-db-core/src/main/java/com/softicar/platform/db/core/transaction/IDbTransaction.java:15]: The definition of `logical` seems to differ from that in the sentence before. I suggest "simulated" or "emulated" here.

File: platform-db-core/src/main/java/com/softicar/platform/db/core/transaction/IDbTransaction.java [hunk 2]
```diff
@@ -1,16 +1,61 @@
...
+ * command.</li>
+ * </ul>
+ */
 public interface IDbTransaction extends ITransaction {
 
 	int getId();
 
-	boolean isRootTransaction();
+	/**
+	 * Tests whether this {@link IDbTransaction} is a root transaction or not.
```
Comment (alexander-schmidt) [inline platform-db-core/src/main/java/com/softicar/platform/db/core/transaction/IDbTransaction.java:30]:
```suggestion
	 * Tests whether this {@link IDbTransaction} is a root transaction.
```

File: platform-db-core/src/main/java/com/softicar/platform/db/core/transaction/IDbTransaction.java [hunk 3]
```diff
@@ -1,16 +1,61 @@
...
+	/**
+	 * Asserts that this {@link IDbTransaction} is a root transaction.
+	 *
+	 * @throws SofticarDeveloperException
+	 *             if the assertion fails
+	 */
+	default void assertIsRootTransaction() {
+
+		if (!isRootTransaction()) {
+			throw new SofticarDeveloperException("Unexpected nesting of transactions.");
```
Comment (alexander-schmidt) [inline platform-db-core/src/main/java/com/softicar/platform/db/core/transaction/IDbTransaction.java:48]: We could be more specific here:
```suggestion
			throw new SofticarDeveloperException("Expected a root transaction but encountered a nested transaction.");
```

File: platform-db-core/src/main/java/com/softicar/platform/db/core/transaction/IDbTransaction.java [hunk 4]
```diff
@@ -1,16 +1,61 @@
...
+		}
+	}
+
+	/**
+	 * Tests whether this {@link IDbTransaction} object represents the currently
```
Comment (alexander-schmidt) [inline platform-db-core/src/main/java/com/softicar/platform/db/core/transaction/IDbTransaction.java:53]:
```suggestion
	 * Tests whether this {@link IDbTransaction} represents the currently
```

Comment (alexander-schmidt) [reply platform-db-core/src/main/java/com/softicar/platform/db/core/transaction/IDbTransaction.java:53]: Also, we should probably clarify what an "active" transaction is.

File: platform-demo-module/src/main/java/com/softicar/platform/demo/module/invoice/DemoInvoiceValidator.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+	protected void validate() {
+
+		if (!isAlphaNumeric(tableRow.getInvoiceNumber())) {
+			addError(AGDemoInvoice.INVOICE_NUMBER, DemoI18n.INVOICE_NUMBERS_MAY_ONLY_CONTAIN_LETTERS_AND_DIGITS);
+		}
+	}
+
+	private static boolean isAlphaNumeric(String text) {
+
+		for (int i = 0; i < text.length(); i++) {
```
Comment (alexander-schmidt) [inline platform-demo-module/src/main/java/com/softicar/platform/demo/module/invoice/DemoInvoiceValidator.java:18]: How about
```
return text.matches("[a-zA-Z0-9]*");
```

File: platform-dom/src/main/java/com/softicar/platform/dom/event/DomEventHandlerNodeCaller.java [hunk 6]
```diff
@@ -25,7 +25,7 @@ public void call() {
 			} else {
 				event.getType().handleEvent(eventNode, event);
 			}
-		} catch (RuntimeException exception) {
+		} catch (Exception exception) {
```
Comment (alexander-schmidt) [inline platform-dom/src/main/java/com/softicar/platform/dom/event/DomEventHandlerNodeCaller.java:28]: Nice catch.
...
...get it? Hehe.

File: platform-emf/src/main/java/com/softicar/platform/emf/form/EmfForm.java [hunk 7]
```diff
@@ -122,9 +124,25 @@ public void focusFrame() {
 
 	// ------------------------------ behavioral configuration ------------------------------ //
 
+	/**
+	 * Sets a callback {@link Consumer} to be called whenever a new
+	 * {@link IEmfTableRow} was inserted through this {@link EmfForm}.
+	 * <p>
+	 * This callback mechanism is primarily intended for UI updates. It is
+	 * called when the new {@link IDbTableRow} was <b>persistently</b> inserted
+	 * and all transactions were successfully <b>committed</b>.
```
Comment (alexander-schmidt) [inline platform-emf/src/main/java/com/softicar/platform/emf/form/EmfForm.java:133]: I think we should be more explicit here, by adding sth. like:
`The callback should <b>not</b> write to the database.`
Then, in the paragraph below, we (already) explain what to do _instead_.

File: platform-emf/src/main/java/com/softicar/platform/emf/form/EmfForm.java [hunk 8]
```diff
@@ -122,9 +124,25 @@ public void focusFrame() {
 
 	// ------------------------------ behavioral configuration ------------------------------ //
 
+	/**
+	 * Sets a callback {@link Consumer} to be called whenever a new
+	 * {@link IEmfTableRow} was inserted through this {@link EmfForm}.
+	 * <p>
+	 * This callback mechanism is primarily intended for UI updates. It is
+	 * called when the new {@link IDbTableRow} was <b>persistently</b> inserted
```
Comment (alexander-schmidt) [inline platform-emf/src/main/java/com/softicar/platform/emf/form/EmfForm.java:132]: We could add some emphasis here, like:
```suggestion
	 * called when the new {@link IDbTableRow} was already <b>persistently</b> inserted
```

File: platform-emf/src/main/java/com/softicar/platform/emf/form/EmfForm.java [hunk 9]
```diff
@@ -122,9 +124,25 @@ public void focusFrame() {
...
+	 * and all transactions were successfully <b>committed</b>.
+	 * <p>
+	 * If you intend to insert additional {@link IEmfTableRow} objects through
```
Comment (alexander-schmidt) [inline platform-emf/src/main/java/com/softicar/platform/emf/form/EmfForm.java:135]: Only inserting is mentioned here - but manipulation of existing records could be required as well. Should we express that?

File: platform-emf/src/main/java/com/softicar/platform/emf/form/EmfForm.java [hunk 10]
```diff
@@ -122,9 +124,25 @@ public void focusFrame() {
...
+	 * the callback, you should use an {@link IEmfSaveHook} instead. Only that
+	 * will ensure that the original {@link IDbTableRow} and the additional
+	 * {@link IDbTableRow} objects will be inserted with the same transaction.
+	 *
+	 * @param callbackAfterCreation
+	 *            the callback {@link Consumer} object (never <i>null</i>)
```
Comment (alexander-schmidt) [inline platform-emf/src/main/java/com/softicar/platform/emf/form/EmfForm.java:141]:
```suggestion
	 *            the callback {@link Consumer} (never <i>null</i>)
```

File: platform-emf/src/main/java/com/softicar/platform/emf/form/EmfFormSaveOrCancelActionsInput.java [hunk 11]
```diff
@@ -53,23 +53,16 @@ public EmfFormSaveOrCancelActionsInput(IEmfFormBody<R> formBody, EmfFormAttribut
 	private void save(boolean closeAfterSave) {
 
 		boolean creation = formBody.getTableRow().impermanent();
-		try (DbTransaction transaction = new DbTransaction()) {
+		try (DbTransaction rootTransaction = new DbTransaction()) {
```
Comment (alexander-schmidt) [inline platform-emf/src/main/java/com/softicar/platform/emf/form/EmfFormSaveOrCancelActionsInput.java:56]: Totally optional but I kind of like this to reduce redundancy:
```suggestion
		try (var rootTransaction = new DbTransaction()) {
```

File: platform-emf/src/main/java/com/softicar/platform/emf/form/popup/EmfFormPopup.java [hunk 12]
```diff
@@ -57,6 +57,9 @@ public R getTableRow() {
 		return form.getTableRow();
 	}
 
+	/**
+	 * Delegates to {@link EmfForm#setCallbackAfterCreation(Consumer)}.
```
Comment (alexander-schmidt) [inline platform-emf/src/main/java/com/softicar/platform/emf/form/popup/EmfFormPopup.java:61]: Why would we explicitly describe this - while we don't do the same for the method below? (`addAdditionalValidator`)

File: platform-emf/src/test/java/com/softicar/platform/emf/form/EmfFormTest.java [hunk 13]
```diff
@@ -29,15 +29,44 @@ public void testCreationOfNewEntity() {
...
+		enterNameAndDayAndClickSaveAndClose(SOME_NAME, SOME_DAY);
+
+		// verify entity
+		assertFalse(entity.impermanent());
+		assertNotNull(entity.getId());
+		assertEquals(SOME_NAME, entity.getName());
+		assertEquals(SOME_DAY, entity.getSimpleEntity().getDay());
+
+		// verify UI
+		findBody().assertContainsText("EXCEPTION FROM CALLBACK");
```
Comment (alexander-schmidt) [inline platform-emf/src/test/java/com/softicar/platform/emf/form/EmfFormTest.java:66]: It's great to finally have a test for this.

**Review by alexander-schmidt — CHANGES_REQUESTED** [review]

File: platform-db-core/src/main/java/com/softicar/platform/db/core/transaction/IDbTransaction.java (hunk 1 above)
Comment (oliver-richers) [reply platform-db-core/src/main/java/com/softicar/platform/db/core/transaction/IDbTransaction.java:15]: changed text

File: platform-demo-module/src/main/java/com/softicar/platform/demo/module/invoice/DemoInvoiceValidator.java (hunk 5 above)
Comment (oliver-richers) [reply platform-demo-module/src/main/java/com/softicar/platform/demo/module/invoice/DemoInvoiceValidator.java:18]: better :)

File: platform-dom/src/main/java/com/softicar/platform/dom/event/DomEventHandlerNodeCaller.java (hunk 6 above)
Comment (oliver-richers) [reply platform-dom/src/main/java/com/softicar/platform/dom/event/DomEventHandlerNodeCaller.java:28]: exceptional! :P

File: platform-emf/src/main/java/com/softicar/platform/emf/form/EmfForm.java (hunk 7 above)
Comment (oliver-richers) [reply platform-emf/src/main/java/com/softicar/platform/emf/form/EmfForm.java:133]: done

File: platform-emf/src/main/java/com/softicar/platform/emf/form/EmfForm.java (hunk 9 above)
Comment (oliver-richers) [reply platform-emf/src/main/java/com/softicar/platform/emf/form/EmfForm.java:135]: yes, i did so now

File: platform-emf/src/main/java/com/softicar/platform/emf/form/popup/EmfFormPopup.java (hunk 12 above)
Comment (oliver-richers) [reply platform-emf/src/main/java/com/softicar/platform/emf/form/popup/EmfFormPopup.java:61]: i think it's necessary to point the dev to the Javadoc, i did the same for the next 2 methods

File: platform-emf/src/test/java/com/softicar/platform/emf/form/EmfFormTest.java (hunk 13 above)
Comment (oliver-richers) [reply platform-emf/src/test/java/com/softicar/platform/emf/form/EmfFormTest.java:66]: it's exceptional again! :)

**Review by alexander-schmidt — APPROVED** [review]

**Comment by oliver-richers** [comment]
btw, in EAS, this is now shown
image

**Pull request merged**